src/ressa/index.rs

```diff
@@ -0,0 +1,166 @@
+use crate::ressa::explorer::RessaNodeExplorer;
+use crate::Language;
+use bitmaps::Bitmap;
+use derive_new::new;
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::ops::BitOrAssign;
+
+use super::NodePattern;
+
+/// Single indexable reference
+pub type IndexableEntry<'a> = &'a dyn Indexable;
+
+#[derive(Hash, PartialEq, Eq)]
+pub struct IndexableKey {
+    wrapped: String,
+}
+impl IndexableKey {
+    fn new<'a>(entry: IndexableEntry<'a>) -> IndexableKey {
+        IndexableKey {
+            wrapped: format!("{:p}", entry),
+        }
+    }
+}
+
+/// Structure used to index the LAAST
+#[derive(new)]
+pub struct LaastIndex<'a> {
+    /// Index of language to root LAAST node (largest subtrees containing a given language)
+    language_index: HashMap<Language, Vec<IndexableEntry<'a>>>,
+
+    /// Reverse index from AST to language set, adding knowledge of subnode languages to a LAAST node
+    ast_languages: HashMap<IndexableKey, LanguageSet>,
+}
+impl<'a> LaastIndex<'a> {
+    /// Get indexed node for a language
+    pub fn get_roots(&self, language: &Language) -> Option<&Vec<IndexableEntry<'a>>> {
+        self.language_index.get(language)
+    }
+
+    /// Retrieve the language of a subtree
+    pub fn language_in_subtree(&self, language: &Language, subtree: IndexableEntry) -> bool {
+        // Edge case: any language requested
+        if *language == Language::Unknown {
+            return true;
+        }
+
+        // Handle main case
+        // println!("Check index...");
+        match self.ast_languages.get(&IndexableKey::new(subtree)) {
+            Some(descendent_languages) => langset_contains(descendent_languages, language),
+            None => panic!("Unknown subtree provided"),
+        }
+    }
+
+    /// Indexes the given node, if its language is indexed on
+    fn add_if_valid(&mut self, node: IndexableEntry<'a>) {
+        // Index under specific language
+        if let Some(index) = self.language_index.get_mut(&node.get_language()) {
+            // println!("{:#?}", node.get_language());
+            index.push(node);
+        }
+
+        // Index under catchall
+        if let Some(index) = self.language_index.get_mut(&Language::Unknown) {
+            // println!("Catchall");
+            index.push(node);
+        }
+    }
+
+    /// Indexes the given node, if its language is indexed on
+    fn record_subtree_langs(&mut self, node: IndexableEntry<'a>, langs: LanguageSet) {
+        self.ast_languages.insert(IndexableKey::new(node), langs);
+    }
+}
+
+/// Descriptive alias. Also, centralizes size; may need to expand in the future.
+pub type LanguageSet = Bitmap<32>;
+
+/// Retrieve whether a language is recorded
+pub fn langset_contains(langset: &LanguageSet, lang: &Language) -> bool {
+    langset.get(Language::get_index(lang))
+}
+
+/// Record a language to the set
+pub fn langset_set(langset: &mut LanguageSet, lang: &Language) -> bool {
+    langset.set(Language::get_index(lang), true)
+}
```

Comment:
I'm guessing you probably did this as regular functions since you can't implement methods on an external type. Just for future reference or if you want to change it you can encapsulate it with the [newtype pattern](https://rust-unofficial.github.io/patterns/patterns/behavioural/newtype.html)

```rust
pub struct LanguageSet(Bitmap<32>);
impl LanguageSet {
    pub fn contains(&self, lang: Language) -> bool {
        self.0.get(Language::get_index(&lang))
    }
    //...
}
```

You can also actually use pattern matching in function parameters but it doesn't seem to actually work for recognizing a method as a method instead of a function so you'd have to call it like this
```rust
pub struct LanguageSet(Bitmap<32>);
impl LanguageSet {
    pub fn contains(&Self(langset): &Self, lang: Language) -> bool {
        langset.get(Language::get_index(&lang))
    }
    //...
}

LanguageSet::contains(&langset, Language::Unknown);
```

Also, `Language` is bitwise copy so we don't really need a reference here (or in the other places it's passed as a reference).

Reply:
Thank you for showing me this, updated. The solution I made originally felt like C code instead of Rust, and I'm not a fan of writing unidiomatic code like that. Besides, that C idiom is one I'm not a fan of anyway.